src/permission/rule-application.js

```diff
@@ -0,0 +1,420 @@
+		data = JSON.parse( JSON.stringify( currentData ) );
+		jsonPath.setValue( data, newData );
+		return data;
+	} else {
+		this._loadRecord( this._params.name );
+	}
+};
+
+/**
+ * Returns or loads the record's previous value. Only supported for record
+ * write and read operations
+ *
+ * If getData encounters an error, the rule application might already be destroyed
+ * at this point
+ *
+ * @private
+ * @returns {void}
+ */
+RuleApplication.prototype._getOldData = function() {
+	if( this._isDestroyed === true || this._params.rule.hasOldData === false ) {
+		return null;
+	} else if( this._recordData[ this._params.name ] ) {
+		return this._recordData[ this._params.name ];
+	} else {
+		this._loadRecord( this._params.name );
+	}
+};
+
+/**
+ * Compile the list of arguments that will be injected
+ * into the permission function. This method is called
+ * everytime the permission is run. This allows it to merge
+ * patches and update the now timestamp
+ *
+ * @private
+ * @returns {void}
+ */
+RuleApplication.prototype._getArguments = function() {
+	return [
+		this._crossReferenceFn,
+		this._user,
+		this._getCurrentData(),
+		this._getOldData(),
+		Date.now(),
+		this._params ? this._params.action : null
+	].concat( this._pathVars );
+};
+
+/**
+ * Returns the data for the user variable. This is only done once
+ * per rule as the user is not expected to change
+ *
+ * @private
+ * @returns {void}
+ */
+RuleApplication.prototype._getUser = function() {
+	return {
+		isAuthenticated: this._params.username !== OPEN,
+		id: this._params.username,
+		data: this._params.authData
+	};
+};
+
+/**
+ * Applies the compiled regexp for the path and extracts
+ * the variables that will be made available as $variableName
+ * within the rule
+ *
+ * This is only done once per rule as the path is not expected
+ * to change
+ *
+ * @private
+ * @returns {Array} pathVars
+ */
+RuleApplication.prototype._getPathVars = function() {
+	return this._params.name.match( this._params.regexp ).slice( 1 );
+};
+
+/**
+ * Returns true if all loading operations that are in progress have finished
+ * and no run has been scheduled yet
+ *
+ * @private
+ * @returns {Boolean}
+ */
+RuleApplication.prototype._isReady = function() {
+	var isLoading = false;
+
+	for( var key in this._recordData ) {
+		if( this._recordData[ key ] === LOADING ) {
+			isLoading = true;
+		}
+	}
+
+	return isLoading === false && this._runScheduled === false;
+};
+
+/**
+ * Loads a record with a given name. This will either result in
+ * a onLoadComplete or onLoadError call. This method should only
+ * be called if the record is not already being loaded or present,
+ * but I'll leave the additional safeguards in until absolutely sure.
+ *
+ * @param   {String} recordName
+ *
+ * @private
+ * @returns {void}
+ */
+RuleApplication.prototype._loadRecord = function( recordName ) {
+	/* istanbul ignore next */
+	if( this._recordData[ recordName ] === LOADING ) {
+		return;
```

Comment:
worth logging to see if this is ever reached? No way to find out otherwise ( Pretty useless logline though )